### PR TITLE
SAK-46686 Reimplementing showCalculatedGrade=false

### PR DESCRIPTION
**Pull request by jesusmmp**
https://sakaiproject.atlassian.net/browse/SAK-46686
https://github.com/sakaiproject/sakai-reference/pull/145

New implementation for this one that was reverted here:
https://github.com/sakaiproject/sakai/pull/10070

### CODE REVIEW

File: edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java [hunk 1]
```diff
@@ -3223,11 +3222,21 @@ public Object doInHibernate(final Session session) throws HibernateException {
 				// set date recorded
 				cg.setDateRecorded(gr.getDateRecorded());
 
+				// set entered points
+				cg.setEnteredPoints(gr.getEnteredPoints());
```
Comment (ottenhoff) [inline edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java:3226]: This is still setting the record in the database?

**Comment by ottenhoff** [comment]
I haven't tested this code, but it doesn't look right to me. The student's grade should not be manipulated in edu-services.

If the goal is to show an equivalent number of points while in the Gradebook, then the code should only be in GradebookNG. You should look for the setting, look for an equivalent number of points, and display it. I wouldn't expect that code to be more than 20 lines in 2 files.

I am firmly against large changes here. Making large changes to edu-services is a recipe for bugs and regressions.

File: edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java (hunk 1 above)
Comment (jesusmmp) [reply edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java:3226]: This is a new column on DB. I'll do another PR on the scripts.

**Comment by ottenhoff** [comment]
You should not be touching the database at all. If you want the display to show "C (73)" because of an override, this can all be calculated without manipulating the database.

**Comment by jesusmmp** [comment]
@ottenhoff You are wrong. It's really imposible to show the points without saving it on DB.

**Comment by ottenhoff** [comment]
> It's really imposible to show the points without saving it on DB.

How does this make any sense? You have a grade override ("C"). You have a mapping. Why would you need a database insert?

**Comment by jesusmmp** [comment]
Let's see an example: That grade was override with a 92.

USA BEHAVIOUR
image

How can I recover that 92? It's really imposible if you don't save it

**Comment by jesusmmp** [comment]
SPANISH BEHAVIOUR
image

**Comment by ottenhoff** [comment]
Absolutely, if you override with a 92 you should save 92 and nothing else. If you override with a "C", you should save "C" and nothing else.

**Comment by jesusmmp** [comment]
I think it's not a good idea to mix scale letters and points on the same column on DB.

**Comment by jesusmmp** [comment]
And we need to change all the log behaviour because there always appears override with letter although you override with points.

image

**Comment by ottenhoff** [comment]
Here is everything I will test for before being comfortable with this PR:

* Percentage grading works fine with grade override
* Points display in Course Grade Override Panel when using "show points" should not change after this patch
* Points display when using show points on the Course Grade column should not change after this patch
* Calculated grade when using custom export with "show points" enabled and selected points total for grade should not change after this patch
* Can grade override using exact points or letter grade
* If override with points e.g., "92.5" this is the exact number shown to instructor and student as student grade
* If override with grade letter ("C"), the calculated grade is still shown (by default)

**Comment by jesusmmp** [comment]
Thanks @ottenhoff. One thing:

- If override with points e.g., "92.5" this is the exact number shown to instructor and student as student grade.
I think your behaviour is that the calculated grade is still shown (by default)

so showing the override points only works with: gradebook.coursegrade.showCalculatedGrade=false.

**Comment by jesusmmp** [comment]
@ottenhoff Have you tested it? Any problem?

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/CourseGradeOverridePanel.java [hunk 2]
```diff
@@ -123,18 +125,23 @@ public void onSubmit(final AjaxRequestTarget target, final Form<?> form) {
 					
 					if (!schema.containsKey(newGrade)) {
 						try {
-							newGrade = getGradeFromNumber(newGrade, schema, currentUserLocale);
+							final Locale englishLocale = new Locale("en", "US");
```
Comment (ottenhoff) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/CourseGradeOverridePanel.java:128]: is this englishLocale used anywhere?

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/CourseGradeOverridePanel.java [hunk 3]
```diff
@@ -251,17 +259,54 @@ private String getGradeFromNumber(String newGrade, Map<String, Double> schema, L
...
 	}
 
+	/**
+	 *
+	 * @param gradeScale the scale value to convert
+	 * @param schema the current schema of Gradebook
+	 * @param currentUserLocale the locale to format the grade with the right decimal separator
+	 * @return the grade
+	 */
+	protected static String getNumberFromGrade(String gradeScale, Map<String, Double> schema, Locale currentUserLocale) {
```
Comment (ottenhoff) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/CourseGradeOverridePanel.java:282]: can we put these two methods into a utility or helper class so they can be tested and re-used?

**Comment by ottenhoff** [comment]
My testing went well on all cases

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/CourseGradeOverridePanel.java (hunk 3 above)
Comment (jesusmmp) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/CourseGradeOverridePanel.java:282]: Where should I put them? On https://github.com/sakaiproject/sakai/tree/master/kernel/kernel-util/src/main/java/org/sakaiproject/util?

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/CourseGradeOverridePanel.java (hunk 2 above)
Comment (jesusmmp) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/CourseGradeOverridePanel.java]: Removed.

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/CourseGradeOverridePanel.java (hunk 3 above)
Comment (ottenhoff) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/CourseGradeOverridePanel.java:281]: FormatHelper.java?

Comment (jesusmmp) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/CourseGradeOverridePanel.java]: Done. Should I change the name: TestCourseGradeOverridePanel.java? What name it should be? TestFormatHelper.java?

**Review by ottenhoff — APPROVED** [review]

**Pull request merged**